docs: poc document for consistent batching configuration in log agent (#3402)
kyma-project/telemetry-manager: docs/contributor/pocs/consistent-batching-across-components/02-log-agent-batching.md

Reviewer: Please add numbers in front of file names - stanza investigation must be read before the exporterbatcher investigation

This investigation determines the ideal exporter batcher configuration after enabling the `SynchronousLogEmitter`. The [stanza-filelog-batching](01-stanza-filelog-batching.md) investigation recommends this change.

The chosen configuration must satisfy the following criteria:

- **Batch size** must be large enough to avoid sending too many small requests to the backend, but small enough to stay within the backend's maximum payload size.
- **Throughput** must remain the same before and after enabling the `SynchronousLogEmitter`.
- **Resource consumption** must remain the same before and after enabling the `SynchronousLogEmitter`.
- **Backpressure** must propagate back to the receiver.

## Understand the Exporter Batcher Architecture

The exporter batcher uses a pull-based batching model with the following data path:

1. **Queue** - The `Send()` method places individual requests into the queue using `queue.Offer()`.
2. **Consume** - The queue's consumers call the batcher's `Consume()` method to pull requests and accumulate them into batches.
3. **Flush** - The batcher maintains a `currentBatch` and flushes it when the batch reaches the minimum size or the flush timeout expires.

> Requests entering the queue come directly from the previous component in the OTel pipeline. The batcher then re-batches these requests into larger batches on the consumer side.

Reviewer 2: simpler:

```suggestion
Requests enter the queue directly from the previous component in the OTel pipeline. The batcher then re-batches these requests into larger batches on the consumer side.
```

### Configure the Exporter Batcher

The exporter batcher has several configuration parameters, all documented in the [exporterhelper package](https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/exporterhelper).

The `sending_queue` and the `batch` each have an independent `sizer` setting that determines the unit of measurement:

- The `sending_queue` defaults to the `requests` sizer. This means `queue_size` counts the number of individual requests. Each request can contain many log records.
- The `batch` defaults to the `items` sizer. This means `min_size` and `max_size` count individual log records.
- The `bytes` sizer is also available. It measures the serialized byte size of each request. This gives precise control over memory usage and batch payload sizes but requires serializing every request to calculate its size.

> When both sizers are set to the same unit, the `queue_size` must not be smaller than the batch `min_size`.

Reviewer 2: ```suggestion
When both sizers use the same unit, `queue_size` must be greater than or equal to the batch `min_size`.
```

> Even though the `sending_queue` and `batch` can use different sizers, `requests` and `items` respectively, the `queue_size` must still be greater than or equal to the batch `min_size` in practice. The following scenario illustrates why.

Reviewer 2: ```suggestion
Even though the `sending_queue` and `batch` can use different sizers (`requests` and `items` respectively), in practice, `queue_size` must still be greater than or equal to the batch `min_size`. The following scenario illustrates why:
```

The `SynchronousLogEmitter` can produce very small batches, as small as a single log record per request, because each reader emits independently. Consider a worst-case scenario where 1024 Pods each produce 1 log per second. Each log becomes a separate single-record request in the queue.

If `queue_size` is smaller than `min_size`, for example `queue_size: 1000` with `min_size: 1024`, the batcher stalls:

1. The queue fills up to its capacity of `1000` requests, holding 1000 log records with one record per request.
2. The batcher needs `1024` records to reach `min_size` and flush, but the queue can hold only `1000`.
3. The `flush_timeout` has not elapsed yet, so a time-based flush does not trigger either.
4. New requests cannot enter the queue because it is full, and the batcher cannot drain the queue because the batch is not large enough to flush.

> The queue stays fully occupied until the `flush_timeout` expires. At that point, the batcher flushes whatever it has accumulated. During this window, all incoming requests block. Setting `queue_size` to at least `min_size`, in this case `1024`, prevents this bottleneck.

Reviewer 2: shorter

```suggestion
The queue stays fully occupied until the `flush_timeout` expires and the batcher flushes whatever it has accumulated. During this window, all incoming requests block. Setting `queue_size` to at least `min_size`, in this case `1024`, prevents this bottleneck.
```

## Identify Backend Requirements

The backend services impose payload size and rate limits that constrain the batch configuration.

| Backend           | Maximum Payload Size                                            | Rate Limit                                                                                |
|-------------------|-----------------------------------------------------------------|-------------------------------------------------------------------------------------------|
| SAP Cloud Logging | `4 MiB` per request. Requests exceeding this limit return `413` | 100 x 2 KiB logs/s for Standard plan, 1000 x 2 KiB logs/s for Large plan                 |

> | SAP Cloud ALM     | `1 MB` for JSON payloads, `150 KB` for protobuf binary payloads | `100` requests/minute for production plans, `100` requests/24 hours for development plans |

Reviewer 2: ```suggestion
| SAP Cloud ALM     | `1 MB` for JSON payloads, `150 KB` for Protobuf binary payloads | `100` requests/minute for production plans, `100` requests/24 hours for development plans |
```

|---------------------------------------|----------------------|------------------------------------------------------------------------------------|
| `sending_queue::sizer`                | `requests` (default) | Each queue slot holds one request, regardless of how many records it contains.     |
| `sending_queue::queue_size`           | `1024`               | Provides sufficient buffer for incoming batches.                                   |
| `sending_queue::batch::sizer`         | `items` (default)    | The batcher counts individual log records when deciding whether to flush.          |
| `sending_queue::batch::min_size`      | `1024`               | Matches the batch size used by other components.                                   |
| `sending_queue::batch::max_size`      | `1024`               | Matches the batch size used by other components.                                   |
| `sending_queue::batch::flush_timeout` | `10s`                | Matches the flush interval used by other components.                               |

> The same assumptions as the baseline apply, that is, 2 KB per record and up to 100 records per request from the `SynchronousLogEmitter`. The maximum queue memory is:

Reviewer 2: ```suggestion
The same assumptions as the baseline apply: 2 KB per record and up to 100 records per request from the `SynchronousLogEmitter`.

The maximum queue memory is:
```

As the comparison table shows, gRPC calls drop from 1235 to 156, approximately 9x, with the bytes batcher assembling ~2.17 MB per call compared to ~273 KB in the baseline. CPU is slightly lower with the exporter batcher. RSS and `heap_alloc` are slightly higher, which is expected because the batcher holds a larger in-flight buffer to accumulate to its byte threshold before flushing. No records are lost.

Other components currently use `batchprocessor`, which does not support sizers other than `items`. Using sizers other than `items` would require migrating all other components to the exporter batcher. The [batch processor migration ADR](../../arch/029-batch-processor-migration-to-exporterhelper.md) investigates this migration and identifies certain risks and unwanted behavior.

## Conclusions

The batching configuration cannot be consistent across all components because other components use `batchprocessor` with the `items` sizer. The log agent benefits from using the `bytes` sizer to bound queue memory usage and control batch sizes based on payload size. Tests show little to no performance degradation from introducing the exporter batcher.

> The recommended configuration uses the `bytes` sizer to ensure that batches exceeding `4 MB` are split and batches of approximately `2 MB` are flushed:

Reviewer 2: shorter:

```suggestion
The recommended configuration uses the `bytes` sizer to split batches that exceed 4 MB and to flush batches of approximately 2 MB:
```
